pkg/sqlutil/pg/stats.go

```diff
@@ -0,0 +1,132 @@
+package pg
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+)
+
+const dbStatsInternal = 10 * time.Second
+
+var (
+	promDBConnsMax = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "db_conns_max",
+		Help: "Maximum number of open connections to the database.",
+	})
+	promDBConnsOpen = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "db_conns_open",
+		Help: "The number of established connections both in use and idle.",
+	})
+	promDBConnsInUse = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "db_conns_used",
+		Help: "The number of connections currently in use.",
+	})
+	promDBWaitCount = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "db_wait_count",
+		Help: "The total number of connections waited for.",
+	})
+	promDBWaitDuration = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "db_wait_time_seconds",
+		Help: "The total time blocked waiting for a new connection.",
+	})
+)
+
+func publishStats(stats sql.DBStats) {
+	promDBConnsMax.Set(float64(stats.MaxOpenConnections))
+	promDBConnsOpen.Set(float64(stats.OpenConnections))
+	promDBConnsInUse.Set(float64(stats.InUse))
+
+	promDBWaitCount.Set(float64(stats.WaitCount))
+	promDBWaitDuration.Set(stats.WaitDuration.Seconds())
+}
+
+type StatsReporterOpt func(*StatsReporter)
+
+func StatsInterval(d time.Duration) StatsReporterOpt {
+	return func(r *StatsReporter) {
+		r.interval = d
+	}
+}
+
+func StatsCustomReporterFn(fn ReportFn) StatsReporterOpt {
+	return func(r *StatsReporter) {
+		r.reportFn = fn
+	}
+}
+
+type (
+	StatFn   func() sql.DBStats
+	ReportFn func(sql.DBStats)
+)
+
+type StatsReporter struct {
+	statFn   StatFn
+	reportFn ReportFn
+	interval time.Duration
+	cancel   context.CancelFunc
+	lggr     logger.Logger
+	once     sync.Once
+	wg       sync.WaitGroup
+}
+
+func NewStatsReporter(fn StatFn, lggr logger.Logger, opts ...StatsReporterOpt) *StatsReporter {
+	r := &StatsReporter{
+		statFn:   fn,
+		reportFn: publishStats,
+		interval: dbStatsInternal,
+		lggr:     logger.Named(lggr, "StatsReporter"),
+	}
+
+	for _, opt := range opts {
+		opt(r)
+	}
+
+	return r
+}
+
+func (r *StatsReporter) Start() {
+	startOnce := func() {
+		r.wg.Add(1)
+		r.lggr.Debug("Starting DB stat reporter")
+		ctx, cancelFunc := context.WithCancel(context.Background())
```

Comment:
should this be a context.Background? is there an overarching resource lifecycle that would make sense to drill through here?

Reply:
It would still be a `context.Background()` if we wired up a `StopChan`. But since there is just one routine, that complexity is not necessary.

Reply:
Cancel and Wait() still happen normally